File: Heap and Priority Queue/p347_top_k_frequent_elements.cpp

class Solution {
public:
  vector<int> topKFrequent(vector<int> &nums, int k) {
    // count it in a map
    vector<int> ans = {};
    unordered_map<int, int> map;

    // key= number, val= frequency
    for (int i = 0; i < nums.size(); i++) {
      map[nums[i]] += 1;
    }

    // append key values to priority queue (heap)
    priority_queue<pair<int, int>> pq;
    for (const auto &[num, freq] : map) {
      pq.push({freq, num});
    }
    // pop k times from heap
    for (int i = 0; i < k; i++) {
      auto top = pq.top();
      pq.pop();
      ans.push_back(top.second);
    }

    return ans;
  }
};

File: Heap and Priority Queue/p347_top_k_frequent_elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        import heapq

        heap=[]
        res=[]
        hashmap= defaultdict(int)
        for num in nums:
            hashmap[num]+=1
        print(hashmap)

        heapq.heapify(heap)
        for key,val in hashmap.items():
            heapq.heappush(heap,[-val,key])

        print("heap is",heap)
        for i in range(k):
            pair= heapq.heappop(heap)
            res.append(pair[1])

        return res

        
